File: backoffice/src/shared/infrastructure/events/DomainEventsLocator.py
"""
 *
 * Libraries 
 *
"""

import inspect
from types                  import ModuleType
from src.shared.domain      import DomainEvent
from ..ContextModulesFinder import ContextModulesFinder

"""
 *
 * Class
 *
"""

class DomainEventsLocator:

    """
     *
     * Methods 
     *
    """

    @classmethod
    def locate( cls ) -> list[type]:
        # Variables
        eventClasses     : list[type]
        domainModules    : list[ModuleType]
        satisfiedClasses : list[type]        
        # Code
        domainModules = ContextModulesFinder.findDomainModules()
        eventClasses  = []
        for domainModule in domainModules:
            satisfiedClasses = list( map( lambda members: members[1], inspect.getmembers( 
                domainModule, 
                cls.__isEventClass                    
            ) ) )           
            eventClasses.extend( satisfiedClasses )
        return eventClasses
    
    @classmethod
    def __isEventClass( cls, anyClass : object ) -> bool:
        return inspect.isclass( anyClass ) and \
            issubclass( anyClass, DomainEvent ) and \
            anyClass != DomainEvent

File: preparation/src/shared/infrastructure/events/DomainEventsLocator.cs
/* 
 * 
 * Libraries
 *
*/

using System.Reflection;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class DomainEventsLocator { 

        /* 
         * 
         * Methods
         *
        */

        public static List<Type> locate() {
            // Variables
            List<Type>            eventClasses;
            Assembly              assembly;    
            IEnumerable<TypeInfo> satisfiedClasses;
            IEnumerable<TypeInfo> classTypes;
            // Code
            assembly   = Assembly.Load( "preparation" );
            classTypes = assembly.ExportedTypes.Select(
                i => i.GetTypeInfo()
            ).Where( i => i.IsClass && !i.IsAbstract );
            satisfiedClasses = classTypes.Where( i => i.IsSubclassOf( typeof( DomainEvent ) ) );
            eventClasses     = new List<Type>();
            eventClasses.AddRange( satisfiedClasses.Select( i => i.AsType() ) );
            return eventClasses;
        }
    
    }

}